formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/schema/Generation.kt

```diff
@@ -0,0 +1,413 @@
+        }
+
+        val fieldDescriptor = messageDescriptor.getElementDescriptor(index)
+
+        nestedTypes += when {
+            fieldDescriptor.isProtobufNamedType -> generateNamedType(messageDescriptor, index, builder)
+            fieldDescriptor.isProtobufRepeated -> generateListType(messageDescriptor, index, builder)
+            fieldDescriptor.isProtobufMap -> generateMapType(messageDescriptor, index, builder)
+            else -> throw IllegalStateException(
+                "Unprocessed message field type with serial name " +
+                        "'${fieldDescriptor.serialName}' and kind '${fieldDescriptor.kind}'"
+            )
+        }
+
+
+        val annotations = messageDescriptor.getElementAnnotations(index)
+        val number = annotations.filterIsInstance<ProtoNumber>().singleOrNull()?.number ?: index + 1
+        if (!usedNumbers.add(number)) {
+            throw IllegalArgumentException("Field number $number is repeated in the class with serial name ${messageDescriptor.serialName}")
+        }
+
+        builder.append(' ')
+        builder.append(fieldName)
+        builder.append(" = ")
+        builder.append(number)
+        builder.appendLine(';')
+    }
+    builder.appendLine('}')
+
+    return nestedTypes
+}
+
+private fun generateNamedType(
+    messageDescriptor: SerialDescriptor,
+    index: Int,
+    builder: StringBuilder
+): List<SerialDescriptor> {
+    val fieldDescriptor = messageDescriptor.getElementDescriptor(index)
+    val nestedTypes: List<SerialDescriptor>
+    val typeName: String = when {
+        messageDescriptor.isSealedPolymorphic && index == 1 -> {
+            builder.appendLine("  // decoded as message with one of these types:")
+            nestedTypes = fieldDescriptor.elementDescriptors.toList()
+            nestedTypes.forEachIndexed { _, childDescriptor ->
+                builder.append("  //   message ").append(childDescriptor.messageOrEnumName)
+                    .append(", serial name '").append(removeLineBreaks(childDescriptor.serialName)).appendLine('\'')
+            }
+            fieldDescriptor.scalarTypeName()
+        }
+        fieldDescriptor.isProtobufScalar -> {
+            nestedTypes = emptyList()
+            fieldDescriptor.scalarTypeName(messageDescriptor.getElementAnnotations(index))
+        }
+        fieldDescriptor.isOpenPolymorphic -> {
+            nestedTypes = listOf(SyntheticPolymorphicDescriptor)
+            SyntheticPolymorphicDescriptor.serialName
+        }
+        else -> {
+            // enum or regular message
+            nestedTypes = listOf(fieldDescriptor)
+            fieldDescriptor.messageOrEnumName
+        }
+    }
+
+    if (fieldDescriptor.isNullable && messageDescriptor.isElementOptional(index)) {
+        builder.appendLine("  // WARNING: this field is nullable and has default value, it's impossible to unambiguously interpret an absence value.")
```

Comment:
This bigger message bears the same information that the following smaller one: default value is missing from schema. 't's impossible to unambiguously interpret an absence value' is a valid point, but does not tell user what happens because of this. Also most of the default values would be 'null' and 'null value does not support' may confuse users (because we actually support it in that case)

I'd rather remove this clause and join it with the following with simplified a message to e.g. 'Absence of this value is decoded to a default value declared in the Kotlin class and omitted from this schema'. Additional restrictions and null handling details should be described in the documentation.